executions/graphql-kotlin-transaction-batcher-instrumentation/src/main/kotlin/com/expediagroup/graphql/transactionbatcher/instrumentation/extensions/ConcurrentHashMapExtensions.kt

```diff
@@ -0,0 +1,34 @@
+/*
+ * Copyright 2022 Expedia, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.expediagroup.graphql.transactionbatcher.instrumentation.extensions
+
+import java.util.concurrent.ConcurrentHashMap
+
+/**
+ * if [ConcurrentHashMap] contains a given [key] execute a given [block] function while
+ * holding the monitor of the given object lock
+ *
+ * @param key to check if there is a value associated with
+ * @param block function to execute over the associated value while holding his monitor
+ * @return result of [block] function that was executed over the associated values
+ */
+fun <K, V, R> ConcurrentHashMap<K, V>.synchronizeIfPresent(key: K, block: (V) -> R): R? =
```

Comment:
this definitely shouldn't be a public API - maybe we should make it private and move it under `ExecutionLevelInstrumentationState`?

Reply:
what about `internal` as the other extension functions ?

Reply:
was just wondering - do we even need this? why not just use https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfPresent-K-java.util.function.BiFunction-? which should be thread safe?

Reply:
`computeIfPresent` and `computeIfAbsent` apply a lock on the whole map.

As documentation indicates: 
```
Some attempted update operations on this map by other threads may be blocked while computation is in progress
```
and we are interested on applying a lock on the monitor of the Value as it would be possible to have parallelism altering the concurrentHashMap but we need to be atomic at the value level

Reply:
having this extension function will allow to concurrently modify this map without locking it, but actually applying a lock over the value that we are attempting to modify

Reply:
[graphql-java](https://github.com/graphql-java/graphql-java/blob/master/src/main/java/graphql/execution/instrumentation/dataloader/FieldLevelTrackingApproach.java#L127) synchronizes the access to the executionState, for the possible parallelism reason

Reply:
IMHO you are preemptively attempting to optimize thing that most likely doesn't matter and introduces unncessary complexity - you are just increasing the counts in the `ExecutionBatchState`. Default behavior from `computeIfPresent`/`computeIfAbsent` should be good enough and we should optimize it if there is actually a perf problem with it.